src/sentry/snuba/metrics/extraction.py

```diff
@@ -1179,11 +1180,16 @@ def mri(self) -> str:
     @cached_property
     def _query_str_for_hash(self) -> str:
         """Returns a hash of the query and field to be used as a unique identifier for the on-demand metric."""
-        str_to_hash = f"{self._field_for_hash()};{self._query_for_hash()}"
+        if self.spec_version.flags == {"include_environment_tag"}:
+            str_to_hash = f"{self._field_for_hash()};{self._query_for_hash()};{self._tag_for_field('environment')}"
+        else:
+            str_to_hash = f"{self._field_for_hash()};{self._query_for_hash()}"
+
```

Comment:
I feel like this should actually be in `self.groupbys` for the new version, and that should be a computed param based on the passed `groupbys` param to the spec, and optionally add env as a group-by. The new spec version can just have environment automatically added in the groupbys, which also means we can remove the condition in `to_metric_spec`.

Additionally we might want to add a check to avoid the case when you build a widget that already _has_ environment tag (since someone is explicitly grouping by env) in addition to the dynamic tag that's added, a group by doesn't need to be added twice.